File: js/lib/MessageBroker/dispatchers/RabbitMQ/routings/PatternMatching/PatternMatchingRabbitMQTopic.js
const BaseRabbitMQTopic = require("../BaseRabbitMQTopic");
const PatternMatchingRabbitMQRoom = require("./PatternMatchingRabbitMQRoom");

class PatternMatchingRabbitMQTopic extends BaseRabbitMQTopic {
  constructor({ name, channel }) {
    super();

    this.name = name;
    this.channel = channel;
    channel.assertExchange(name, 'topic', {durable: false});
  }

  async createRoom({ name }) {
    let room = await PatternMatchingRabbitMQRoom.create({ 
      name, 
      topicName: this.name, 
      channel: this.channel 
    });

    this._addRoom(room);

    return room;
  }

  publish({ room, payload, options }) {
    this.channel.publish(this.name, room, new Buffer(payload), options);
  }
}

module.exports = PatternMatchingRabbitMQTopic

File: ruby/lib/MessageBroker/dispatchers/RabbitMQ/routings/PatternMatching/PatternMatchingRabbitMQTopic.rb
require_relative '../BaseRabbitMQTopic'
require_relative './PatternMatchingRabbitMQRoom'

class PatternMatchingRabbitMQTopic < BaseRabbitMQTopic
  def initialize name:, channel:
    super()

    @channel = channel
    @exchange = @channel.topic(name)
  end  

  def createRoom name:
    room = PatternMatchingRabbitMQRoom.new(
      name: name,
      channel: @channel,
      exchange: @exchange
    )
    
    addRoom(room: room)

    return room
  end

  def publish room:, payload:, correlationId: nil, replyTo: nil
    @exchange.publish(
      payload, 
      :routing_key => room, 
      :correlation_id => correlationId,
      :replyTo => replyTo
    )
  end
end